models_evaluation.py

```diff
@@ -0,0 +1,176 @@
+    :param models_path: string
+    :param analogies: dict
+    :param first_n: int
+    :return: DataFrame
+    """
+    result = pd.DataFrame(
+        columns=["Model", "Type of questions", "Accuracy1 (for present questions)", "Accuracy2 (for all questions)",
+                 "Number of questions", "Number of present questions", "No words in vocabulary"])
+
+    for file in listdir(models_path):
+        if not isfile(join(models_path, file)):
+            continue
+        name = file.replace(".", "_")
+        logger.info(f"Model {name} loading...")
+
+        # TODO: Ideally here we should check the file extension and load binary fb vectors or textual w2v
+        model = gensim.models.fasttext.load_facebook_vectors(models_path + file)
+
+        logger.info("-------------------------------------------------------")
+        logger.info("Starting evaluation")
+
+        total_analogies = len(list((itertools.chain(*analogies.values()))))
+
+        total_correct = 0
+        present_analogies = 0
+
+        for key, questions in analogies.items():
+            section_correct = 0
+            total_section = len(questions)
+            present_analogies += total_section
+            # TODO: validate how many tasks for each category do we have?
+            for i in range(len(questions)):
+                try:
+                    sim = list(list(zip(*(
+                        model.most_similar(positive=[questions[i][0], questions[i][3]], negative=[questions[i][2]],
+                                           topn=first_n))))[0])
+                    if (questions[i][1] in sim
+                            or questions[i][1].lower() in sim
+                            or questions[i][1] in [s.lower() for s in sim]
+                            or questions[i][1].lower() in [s.lower() for s in sim]):
+                        section_correct += 1
+                        total_correct += 1
+                    else:
+                        try:
+                            # TODO: double check if that contributes any boost to the score on our vectors
+                            #       If not: make it optional and disabled by default
+                            sim = list(list(zip(*(
+                                model.most_similar(positive=[questions[i][0].lower(), questions[i][3].lower()],
+                                                   negative=[questions[i][2].lower()],
+                                                   topn=first_n))))[0])
+
+                            if questions[i][1].lower() in sim or questions[i][1].lower() in [s.lower() for s in sim]:
+                                section_correct += 1
+                                total_correct += 1
+
+                            else:
+                                pass
+
+                        except:
+                            total_section = total_section - 1
+
+                except:
+                    try:
```

Comment:
What worries me is that we have two identical pieces of code (or at least so it seems) right above and below.

One that below fires, when there is an exception (any exception, which is a not very good practice, especially since it's silenced).
Second one fires if case-sensitive search gave no result.

Could we somehow refactor that. For example, raise an exception in the else clause above and that catch it as we do below.

Or adding an extra check to the if above, to cover cases that are raising exceptions.